src/scenic/syntax/compiler.py

```diff
@@ -0,0 +1,1756 @@
+        subRunner = ast.Call(subHandler, subArgs, keywords)
+        return self.generateInvocation(node, subRunner, ast.YieldFrom)
+
+    @context(Context.TOP_LEVEL)
+    def visit_Record(self, node: s.Record):
+        return self.createRequirementLike("record", node.value, node.lineno, node.name)
+
+    @context(Context.TOP_LEVEL)
+    def visit_RecordInitial(self, node: s.RecordInitial):
+        return self.createRequirementLike(
+            "record_initial", node.value, node.lineno, node.name
+        )
+
+    @context(Context.TOP_LEVEL)
+    def visit_RecordFinal(self, node: s.RecordFinal):
+        return self.createRequirementLike(
+            "record_final", node.value, node.lineno, node.name
+        )
+
+    @context(Context.TOP_LEVEL)
+    def visit_TerminateWhen(self, node: s.TerminateWhen):
+        return self.createRequirementLike(
+            "terminate_when", node.cond, node.lineno, node.name
+        )
+
+    @context(Context.TOP_LEVEL)
+    def visit_TerminateSimulationWhen(self, node: s.TerminateSimulationWhen):
+        return self.createRequirementLike(
+            "terminate_simulation_when", node.cond, node.lineno, node.name
+        )
+
+    def createRequirementLike(
+        self,
+        functionName: str,
+        body: ast.AST,
+        lineno: int,
+        name: Optional[str] = None,
+        prob: Optional[float] = None,
+    ):
+        """Create a call to a function that implements requirement-like features, such as `record` and `terminate when`.
+
+        Args:
+            functionName (str): Name of the requirement-like function to call. Its signature must be `(reqId: int, body: () -> bool, lineno: int, name: str | None)`
+            body (ast.AST): AST node to evaluate for checking the condition
+            lineno (int): Line number in the source code
+            name (Optional[str], optional): Optional name for requirements. Defaults to None.
+            prob (Optional[float], optional): Optional probability for requirements. Defaults to None.
+        """
+        propTransformer = PropositionTransformer(self.filename)
+        newBody, self.nextSyntaxId = propTransformer.transform(
+            body, self.nextSyntaxId
+        )
+        newBody = self.visit(newBody)
+        requirementId = self._register_requirement_syntax(body)
+
+        return ast.Expr(
+            value=ast.Call(
+                func=ast.Name(functionName, loadCtx),
+                args=[
+                    ast.Constant(requirementId),  # requirement IDre
+                    newBody,  # body
+                    ast.Constant(lineno),  # line number
+                    ast.Constant(name),  # requirement name
+                ],
+                keywords=[ast.keyword(arg="prob", value=ast.Constant(prob))]
+                if prob is not None
+                else [],
+            )
+        )
+
+    def visit_TerminateAfter(self, node: s.TerminateAfter):
```

Comment:
This and `visit_Simulator` below should have `@context(Context.TOP_LEVEL)`.

Reply:
@shumbo